driver-core/src/main/com/mongodb/internal/async/AsyncRunnable.java

```diff
@@ -0,0 +1,165 @@
+ */
+public interface AsyncRunnable {
+
+    static AsyncRunnable startAsync() {
+        return (c) -> c.onResult(null, null);
+    }
+
+    /**
+     * Must be invoked at end of async chain
+     * @param callback the callback provided by the method the chain is used in
+     */
+    void complete(SingleResultCallback<Void> callback); // NoResultCallback
+
+    /**
+     * Must be invoked at end of async chain
+     * @param runnable the sync code to invoke (under non-exceptional flow)
+     *                 prior to the callback
+     * @param callback the callback provided by the method the chain is used in
+     */
+    default void complete(final Runnable runnable, final SingleResultCallback<Void> callback) {
+        this.complete((r, e) -> {
+            if (e != null) {
+                callback.onResult(null, e);
+                return;
+            }
+            try {
+                runnable.run();
+            } catch (Throwable t) {
+                callback.onResult(null, t);
+                return;
+            }
+            callback.onResult(null, null);
+        });
+    }
+
+    /**
+     * See {@link #complete(Runnable, SingleResultCallback)}, but the runnable
+     * will always be executed, including on the exceptional path.
+     * @param runnable the runnable
+     * @param callback the callback
+     */
+    default void completeAlways(final Runnable runnable, final SingleResultCallback<Void> callback) {
+        this.complete((r, e) -> {
+            try {
+                runnable.run();
+            } catch (Throwable t) {
+                callback.onResult(null, t);
+                return;
+            }
+            callback.onResult(r, e);
+        });
+    }
+
+    /**
+     * @param runnable The async runnable to run after this one
+     * @return the composition of this and the runnable
+     */
+    default AsyncRunnable run(final AsyncRunnable runnable) {
+        return (c) -> {
+            this.complete((r, e) -> {
+                if (e != null) {
+                    c.onResult(null, e);
+                    return;
+                }
+                try {
+                    runnable.complete(c);
+                } catch (Throwable t) {
+                    c.onResult(null, t);
+                }
+            });
+        };
+    }
+
+    /**
+     * @param supplier The supplier to supply using after this runnable.
+     * @return the composition of this runnable and the supplier
+     * @param <T> The return type of the supplier
+     */
+    default <T> AsyncSupplier<T> supply(final AsyncSupplier<T> supplier) {
+        return (c) -> {
+            this.complete((r, e) -> {
+                if (e != null) {
+                    c.onResult(null, e);
+                    return;
+                }
+                try {
+                    supplier.complete(c);
+                } catch (Throwable t) {
+                    c.onResult(null, t);
+                }
```

Comment:
I think there are two points:

1. Ensuring that anything thrown by `this` is passed into the callback. Throwing from a callback-method is usually a contract violation, but we claimed that as long as we never called the "unsafe" functional abstract method directly, all throwables (including assertion errors) could be handled via a wrapper in `finish`. This was convenient, and allowed a closer correspondence of sync and async code. If something is thrown, we can presume that a callback has not been invoked, unless the error arose from the callback.
2. Ensuring that anything thrown by the callback is not passed back into the callback. This is a contract violation. It can be caught by wrapping the callback in `finish`, and ignored in the outer try-catch. I don't like that it goes unreported.

We can ensure the error is reported by not ignoring it. This will pass it into the callback, but it will take the exceptional path, so only error handlers (including finally-like constructs) would be invoked. This might result the error being recorded, but also in invalid resource closures and ensuing (ignored) exceptions. What do you think of this?

Reply:
>  It can be caught by wrapping the callback in finish, and ignored in the outer try-catch. I don't like that it goes unreported.

Me neither, I think it should not be ignored. Everything that can't be properly reported, should be turned into an `AssertionError` (unless it is already of this type) and thrown via the stack. `AsyncCallbackFunction.apply` (and, therefore, abstract methods on the rest of async interfaces, as they refer to this one) is already documented as

```java
* @throws RuntimeException Never. Exceptions must be relayed to the {@code callback}.
* @throws Error Never, on the best effort basis. Errors should be relayed to the {@code callback}.
```

The reason the API was documented this way from the beginning is because it is technically impossible to guarantee (i.e., regardless of how perfect the code is) that no `Error`s is ever thrown instead of being passed to a callback.

> pass it into the callback, but it will take the exceptional path, so only error handlers (including finally-like constructs) would be invoked. This might result the error being recorded, but also in invalid resource closures and ensuing (ignored) exceptions. What do you think of this?

If a callback does not release resources on an exceptional path, the callback is incorrect, and there is nothing we can do with that. Since we can't do anything about it, we should not be concerned<sup>1</sup>. Also, while this is a possibility, it's not a guarantee. On the other hand, if we don't call the callback at all, then it is guaranteed that the callback will not release resources it were supposed to, and in addition we lose the error, i.e., this path is worse in all aspects.

<sup>1</sup> We already decided not to be concerned with this when said that instead of executing the action passed to `completeAlways`, we will produce an `AssertionError`.

Reply:
I raised issues specific to the new code where relevant, and could resolve this one. However, I left it unresolved just in case you want to discuss what you wrote above.

Reply:
I think this is fine to resolve.